### PR TITLE
Containerized the ADAM pipeline from toil-scripts (resolves #139)

### LINKED ISSUE #139: Containerize ADAM pipeline
Single node mono-docker for ADAM, similar to https://github.com/BD2KGenomics/cgl-docker-lib/blob/master/rnaseq-cgl-pipeline.

### PR DESCRIPTION
**Pull request by fnothaft**
Provides a single node implementation of the ADAM preprocessing pipeline that is
at https://github.com/BD2KGenomics/toil-scripts/tree/master/src/toil_scripts/adam_pipeline.
Depends on https://github.com/BD2KGenomics/toil-scripts/pull/351.

### CODE REVIEW

File: adam-pipeline/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	wget \
+	apt-transport-https \
+	ca-certificates
+
+# Get the 1.9.1 binary
+RUN wget https://get.docker.com/builds/Linux/x86_64/docker-1.9.0 -O /usr/local/bin/docker
+RUN chmod u+x /usr/local/bin/docker
+
+# Install Toil
+RUN pip install toil
```
Comment (hannes-ucsc) [inline adam-pipeline/Dockerfile]: pin version?

File: adam-pipeline/test/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+*.processed.bam
```
Comment (hannes-ucsc) [inline adam-pipeline/test/.gitignore]: missing NL at EOL

File: adam-pipeline/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+# Install adam pipeline
+RUN git clone https://github.com/fnothaft/toil-scripts /opt/adam-pipeline
+RUN cd /opt/adam-pipeline && git checkout issues/210-adam-local
```
Comment (hannes-ucsc) [inline adam-pipeline/Dockerfile:25]: We mustn't forget that this is temporary and needs to be changed Add GH issue?

File: adam-pipeline/wrapper.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+
+def write_config(mount, args):
+    
+    # open path to write config file at
+    path = "adam-preprocessing-%s.config" % (str(uuid4()))
+    fp = open(path, 'w')
+
+    # generate config
+    config = '''# ADAM preprocessing pipeline configuration file; autogenerated, do not modify!
```
Comment (hannes-ucsc) [inline adam-pipeline/wrapper.py]: Consider textwrap.dedent which is part of Python and lets you indent the body of the multi-line string.

File: adam-pipeline/wrapper.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    fp.close()
+
+    # return config file name
+    return path
+
+
+def call_pipeline(mount, args):
+    
+    # get uuid and make a work directory
+    uuid = 'Toil-Adam-' + str(uuid4())
```
Comment (hannes-ucsc) [inline adam-pipeline/wrapper.py]: I'd prefer `toil-adam-`. I hardly ever use upper-case in snake-case identifiers.

File: adam-pipeline/wrapper.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    if not os.path.isdir(mount + uuid):
```
Comment (hannes-ucsc) [inline adam-pipeline/wrapper.py]: os.path.join

Comment (hannes-ucsc) [inline adam-pipeline/wrapper.py]: probably a case for extracting a local variable to avoid this kind of bug

File: adam-pipeline/wrapper.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+    # Ensure docker.sock is mounted correctly
+    sock_mount = [x['Source'] == x['Destination'] for x in mounts if 'docker.sock' in x['Source']]
+    if len(sock_mount) != 1:
+        raise IllegalArgumentException('Missing socket mount. Requires the following:'
+                                       'docker run -v /var/run/docker.sock:/var/run/docker.sock')
+
+    # Ensure formatting of command for 2 mount points
+    if len(mounts) == 2:
+        if not all(x['Source'] == x['Destination'] for x in mounts):
```
Comment (hannes-ucsc) [inline adam-pipeline/wrapper.py:107]: FYI: I recently added `require()` (proposed by you at some point) to BPL. Not sure if we can assume it being installed as a transitive dependency of toil-scripts. you'd also have to bump the requirement in toil-script's setup.py. Your call.

File: adam-pipeline/wrapper.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+    # generate config
+    config = '''# ADAM preprocessing pipeline configuration file; autogenerated, do not modify!
+num-nodes:
+master-ip:
+dbsnp: %s
+memory: %s
+run-local: true
+local-dir: %s
+''' % (args.known_sites, args.memory, mount)
```
Comment (hannes-ucsc) [inline adam-pipeline/wrapper.py]: Here I would use .format() with named placeholders. Another trick is 

``` py
>>> x="foo"
>>> "{x}".format(**locals())
'foo'
# dot operator works too
>>> "{x} is {x.__class__}".format(**locals())
"foo is <type 'str'>"
```

File: adam-pipeline/dockstore.cwl [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  - id: "#known-sites"
+	type: string
+	description: "Absolute path to the known-sites VCF for masking."
+	inputBinding:
+	  position: 2
+	  prefix: "--known-sites"
+
+  - id: "#memory"
+	type: string
+	description: "Memory in gigabytes to allocate for this pipeline."
```
Comment (hannes-ucsc) [inline adam-pipeline/dockstore.cwl]: GiB or GB?

File: adam-pipeline/README.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+Additionally, we expect a path to a known sites file. This is a file that
+describes positions where variants are known to occur, and is used to mask
+out sites during base quality score recalibration. Typically, a dbSNP VCF file
+is used. This pipeline requires a run environment with at least 10G of memory
+to run BQSR with a known sites file from dbSNP.
+
+This pipeline requires a host with Docker 1.9.0 installed. Other versions of
+Docker will soon be supported.
+
+## Genomic tool containers
```
Comment (hannes-ucsc) [inline adam-pipeline/README.md]: I would remove this section. The commit SHA below is bound to get stale, rendering the WAR ineffective. Can we automate the WAR?

File: adam-pipeline/README.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+```
+
+Note: This is a transient error, so it is likely this step may not be necessary. 
+
+## Running
+
+Mirror the absolute path to the parent directory when using Docker's -v mount
+command. Toil's job store and temporary directories will be created inside this
+mount point. "-v /var/run/docker.sock:/var/run/docker.sock" must always be
+supplied. This is necessary for the ADAM pipeline docker container to run
```
Comment (hannes-ucsc) [inline adam-pipeline/README.md]: Docker

File: adam-pipeline/README.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+the docker containers it relies on.
```
Comment (hannes-ucsc) [inline adam-pipeline/README.md]: Docker

File: adam-pipeline/test.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+                '--known-sites', '/%s/test/small.vcf' % pwd,
+                '--memory', '1']
+        sock = ['-v', '/var/run/docker.sock:/var/run/docker.sock']
+        mirror = ['-v', '/%s/test:/%s/test' % (pwd, pwd)]
+
+        # Check base call for help menu
+        out = subprocess.check_output(base + tool)
+        self.assertTrue('Please see the complete documentation' in out)
+
+        # run full command on sample inputs and check for existance of output file
```
Comment (hannes-ucsc) [inline adam-pipeline/test.py]: existence

**Comment by hannes-ucsc** [comment]
@fnothaft done with review

**Comment by fnothaft** [comment]
Thanks @hannes-ucsc for the quick review! I will address these now.

Some of these issues are duplicated between here and `rnaseq-cgl-pipeline`. I'm going to punt some of those into a follow on ticket, in the interest of time.

File: adam-pipeline/Dockerfile (hunk 3 above)
Comment (fnothaft) [inline adam-pipeline/Dockerfile:25]: Opened #141.

File: adam-pipeline/wrapper.py (hunk 8 above)
Comment (fnothaft) [inline adam-pipeline/wrapper.py]: Unless you've got strong preferences, I prefer this format. LMK.

File: adam-pipeline/wrapper.py (hunk 7 above)
Comment (fnothaft) [inline adam-pipeline/wrapper.py:107]: I've opened #142 for this.

File: adam-pipeline/test/.gitignore (hunk 2 above)
Comment (fnothaft) [inline adam-pipeline/test/.gitignore]: This is addressed.

**Comment by fnothaft** [comment]
Addressed review comments. A few of the comments touched code that the `adam-pipeline` copied from the `rnaseq-cgl-pipeline`. I've split those into a separate commit (8e6358f). LMK if you want that squashed down.

**Pull request merged**